interface/ceed-preconditioning.c

```diff
@@ -1961,6 +1961,105 @@ int CeedOperatorMultigridLevelCreateH1(CeedOperator op_fine, CeedVector p_mult_f
+
+  @ref User
+**/
+int CeedCompositeOperatorScaleMultigridLevel(CeedOperator op_prolong, CeedOperator op_restrict) {
+  CeedCall(CeedOperatorCheckReady(op_prolong));
+  CeedCall(CeedOperatorCheckReady(op_restrict));
+
+  Ceed                ceed;
+  CeedInt             num_sub_ops;
+  CeedSize            mult_l_vec_len;
+  CeedScalar         *mult;
+  CeedVector          mult_l_vec, ones_l_vec;
+  CeedElemRestriction elem_restr;
+  CeedOperator       *sub_ops;
+  CeedOperatorField  *input_fields;
+
+  CeedCall(CeedOperatorGetCeed(op_prolong, &ceed));
+
+  // Get suboperators
+  CeedCall(CeedOperatorGetNumSub(op_prolong, &num_sub_ops));
+  CeedCall(CeedOperatorGetSubList(op_prolong, &sub_ops));
+  if (num_sub_ops == 0) return CEED_ERROR_SUCCESS;
+
+  // Check for matching operators
+  {
+    CeedInt            num_sub_ops_restrict;
+    CeedVector         scale_prolong, scale_restrict;
+    CeedOperator      *sub_ops_restrict;
+    CeedOperatorField *input_fields_restrict;
+
+    // -- Check number suboperators
+    CeedCall(CeedOperatorGetNumSub(op_restrict, &num_sub_ops_restrict));
+    CeedCall(CeedOperatorGetSubList(op_restrict, &sub_ops_restrict));
+    if (num_sub_ops != num_sub_ops_restrict) return CeedError(ceed, CEED_ERROR_INCOMPATIBLE, "Restriction and Prolongation operators must match");
+
+    // -- Check scale vectors
+    CeedCall(CeedOperatorGetFields(sub_ops[0], NULL, &input_fields, NULL, NULL));
+    CeedCall(CeedOperatorGetFields(sub_ops_restrict[0], NULL, &input_fields_restrict, NULL, NULL));
+    CeedCall(CeedOperatorFieldGetVector(input_fields[1], &scale_prolong));
+    CeedCall(CeedOperatorFieldGetVector(input_fields_restrict[1], &scale_restrict));
+    if (scale_prolong != scale_restrict) return CeedError(ceed, CEED_ERROR_INCOMPATIBLE, "Restriction and Prolongation operators must match");
+  }
+
+  // Work vectors
+  CeedCall(CeedOperatorFieldGetElemRestriction(input_fields[1], &elem_restr));
+  CeedCall(CeedElemRestrictionCreateVector(elem_restr, &ones_l_vec, NULL));
+  CeedCall(CeedVectorSetValue(ones_l_vec, 1.0));
+  CeedCall(CeedElemRestrictionCreateVector(elem_restr, &mult_l_vec, NULL));
+  CeedCall(CeedVectorSetValue(mult_l_vec, 0.0));
+  CeedCall(CeedVectorGetLength(mult_l_vec, &mult_l_vec_len));
+  CeedCall(CeedVectorGetArray(mult_l_vec, CEED_MEM_HOST, &mult));
+
+  // Compute multiplicity across suboperators
+  for (CeedInt i = 0; i < num_sub_ops; i++) {
+    const CeedScalar *sub_mult;
+    CeedVector        sub_mult_l_vec, ones_e_vec;
+
+    CeedCall(CeedOperatorGetFields(sub_ops[i], NULL, &input_fields, NULL, NULL));
+    CeedCall(CeedOperatorFieldGetElemRestriction(input_fields[1], &elem_restr));
+    CeedCall(CeedElemRestrictionCreateVector(elem_restr, &sub_mult_l_vec, &ones_e_vec));
+    CeedCall(CeedVectorSetValue(sub_mult_l_vec, 0.0));
+    CeedCall(CeedElemRestrictionApply(elem_restr, CEED_NOTRANSPOSE, ones_l_vec, ones_e_vec, CEED_REQUEST_IMMEDIATE));
```

Comment:
Why not `CeedVectorSetValue(ones_e_vec, 1.0)`?

Reply:
HDiv restrictions prevent this shortcut, right?

Reply:
Hmm, yeah. I remember that now.